### DISCUSSION TITLE
Remote sensor enrolls, then can't authenticate

### DISCUSSION THREAD

**Discussion by jcmadick** [Security-Onion-Solutions/securityonion#2025] [Unsupported Versions]
I've set up a master in AWS, built on Ubuntu 18.04 running the install scripts.  Everything works fine on the master.  I then attempted to set up a remote sensor.  It runs through the setup, then tells me it failed.  Looking at the sosetup.log, I see that the remote sensor authenticates to the master and that the master adds the remote sensor.  However, the next stage, where it checks to see if the salt inion needs to be updated I get the message that "Attempt to authenticate with the salt master failed with timeout error."  Same message is repeated throughout the rest of the log as various actions that require that salt communication fails.  What would cause an initial enrollment to succeed and subsequent attempts to fail?

**Comment by m0duspwnens** [comment]
@jcmadick 
Have you attempted to install the remote node again and see the same result? Can you check the salt log on the master `/opt/so/log/salt/master` and see if there are any clues there? Is there a device between the manager and remote node that could be intercepting the certificates that Salt uses?

**Reply by jcmadick** [reply]
All I see in the master is:

2020-11-19 18:45:50,353 [salt.utils.parsers:1117][WARNING ][3829] Master received a SIGTERM. Exiting.
repeated with only the time changing

This is the second stab creating the remote sensor.  I can't think of anything that would interfere with the certificates. I'll give it another shot and see if I can capture the traffic between the master and node.

**Reply by m0duspwnens** [reply]
Is that message repeating frequently? Is the salt-master service running on the manager? On the manager can you check with `systemctl status salt-master`. Also on the manager, can you verify that the salt-minion key for the remote node is getting accepted? Run `salt-key -l all`

**Reply by jcmadick** [reply]
I'm convinced now it has to do with the certificates.  Salt-master is up and running but nothing showing as accepted on the master.  I'm assuming it has something to do with the external IP vs internal IP (sorry, fairly ignorant in the AWS space currently.)  I can log in via SSH from outside the environment with no issue.  Looking at auth.log, I can see that the soremote user was authenticated during the node setup and executed the salt-key -a for it's key then disconnects.  On the sensor, I see the enrollment process and then the failures after that initial connection.

**Reply by innovate-support** [reply]
I'm having this exact issue except my Manager and Sensor are were built with the latest ISO and are VMware VM's and are both NAT'd through external routers that I've allowed all ports and confirmed nothing is being blocked between Senser and Manger.  I can SSH into Manager from Sensor, and the Salt ports 4505 and 4506 are definitely allowed through the edge firewall because I see the traffic passing when the Sensor tries to authenticate during the setup.  I don't think this is a firewall issue.  

Running systemctl status salt-master on the Manger shows all active and fine.  Running salt-key -l all on Manager only shows the manager key, no minion keys of any kind are listed.

I've re-installed and rebuilt the Manager and Sensor multiple times to attempt to resolve this issue, but Sensor won't authenticate to Manager.  Logs show the "Attempt to authenticate with the salt master failed with timeout error."

**Comment by BustedSec** [comment]
bump

**Comment by innovate-support** [comment]
I can't seem to get a Sensor to connect to a Manager.

Scenario 1:
I've rebuilt my StandAlone SO system to be a distributed setup with a ManagerSearch VM installed on a Dell box with VMware and lots of resources.  I've tried many times to install v2.3.21.iso but am simply unable to complete the install if I leave TheHive enabled.  Seems that many others have the same problem and a solution hasn't manifested yet so I've abandoned that for now.  So after getting the Manager installed it seems to be working and I can log in and see Alerts.  The Manager VM has a 10.11.12.x IP and the edge firewall is NAT'ing to it successfully.  I opened up SSH port 22 and Salt ports 4505-4506 through my edge firewall to allow external Sensors to send their data to the Manager.  All seems fine and functional so far.  Logs look fine too.

When installing a Sensor VM on a different Dell box with VMware that's on a different network I get to the point in the setup where it tries to register it's SSH/Salt key with the Manager, it asks me the hostname and IP of the Manager (which I enter the public IP), and then it drops to terminal so I can enter the 'soremote' credentials.  After entering the password it jumps back to the gui setup screens and continues through the rest of the questions.  When it tries to complete the setup it always fails.  The logs say "Attempt to authenticate with the salt master failed with timeout error." but I can ping and SSH.  When I try to nmap ports 4505-4506 from Sensor to Manager it says 'filtered' which in my experience means that the edge firewall is allowing it through but the destination server isn't responding correctly due to it's own operating system firewall rules.  Running 'systemctl status salt-master' on the Manager shows all active and fine. Running 'sudo salt-key -l all' on Manager only shows the manager key, no minion keys of any kind are listed.  Reviewing discussion boards reveal that others have similar issues and no solution has presented itself yet.

Scenario 2:
Modifying the scenario a little in hopes that things will work, I'm trying something I've had success with before.  I re-installed the ManagerSearch VM using v2.3.2.iso and was able to successfully enable TheHive.  I'm not sure what's changed between v2.3.2 and v2.3.21 but it's broken TheHive install.  After it's final reboot I let it sit and moved onto the Sensor install to see if anything will work with the older iso.  

I re-installed the remote Sensor using v2.3.2.iso also.  After going through all the setup questions and entering the 'soremote' credentials it fails to complete the whole setup and the logs say the same thing after the '95% Verifying Setup' line that it's 'Attempt to authenticate with the salt master failed with timeout error'.  When I try to nmap ports 4505-4506 from Sensor to Manager it says 'filtered' just like before.  I can see traffic successfully passing on those ports through my edge firewall, so I don't think that is the problem.  I've even allowed ALL traffic from source to destination networks to confirm no other mystery port is required.  Tests result in same errors.  And like I said, I can ping, SSH, and HTTP from Minion's network to Manager just fine, but Salt won't connect.  There's something else going on here.  Does Salt have an internal firewall setting that's blocking NAT and only allowing local traffic or something?  Any ideas will be welcomed.

Scenario 3:
Just for kicks I installed a Minion-Test VM on the same box as the Manager and connected to the same virtual internal switch to see if there's some fishy networking going on with NAT.  Result: same error that the Minion timed out and couldn't communicate with the Salt Master.  This is pretty frustrating since supposedly Security Onion was designed to work in a distributed environment, but it's not even working in a 'local' environment.

Scenario 4:
Now I'm gong to try something different.  I've had Sensors working with a Master before, but the Master was defined as a 'Standalone' server instead of a 'ManagerSearch' server.  So, I re-installed the Manager (again) and called it a Standalone system. All installed fine.  Then I re-installed the Sensor on the other network to see if it would connect to the StandAlone Manager.  Failed again, same timeout errors.  I then tried the Minion-Test VM that's on the same virtual internal switch as the the Manager, and it failed too with a timeout error.  I give up.

**Comment by BustedSec** [comment]
It's the firewall. There are ports that need to be open that aren't.

**Reply by innovate-support** [reply]
I'm assuming you are referring to the Security Onion internal firewall. so-allow doesn't have any Salt options, and also doesn't appear to have any options for customization. What's the resolution? Isn't Salt a core function of the system? Isn't it enabled and allowed by default?

**Comment by BustedSec** [comment]
Ok, so this is a just a bandaid  because they closed my issue about fixing
this telling me that it should open these ports during initial setup (even
though it doesn't) but this should get you started. I'm still trying to get
it all working right myself.
sudo iptables --wait -t filter -I INPUT 1 -p tcp --dport 4506 --source
forward_node_ip_here --jump ACCEPT
sudo iptables --wait -t filter -I INPUT 1 -p tcp --dport 4505 --source
forward_node_ip_here  --jump ACCEPT
it won't persist after a reboot

**Comment by BustedSec** [comment]
also, see reference at
https://docs.securityonion.net/en/2.3/firewall.html?highlight=firewall

I've noticed it needs more ports open then it says it does in that
reference doc. Otherwise it connects just fine, but then doesn't send over
any log data.  Here is a netstat from a manager node showing what is
listening from docker-proxy

tcp6 0 0 :::9000 :::* LISTEN 18568/docker-proxy
tcp6 0 0 :::5000 :::* LISTEN 4550/docker-proxy
tcp6 0 0 :::9001 :::* LISTEN 17975/docker-proxy
tcp6 0 0 :::1514 :::* LISTEN 24075/docker-proxy
tcp6 0 0 :::3306 :::* LISTEN 8765/docker-proxy
tcp6 0 0 :::1515 :::* LISTEN 24061/docker-proxy
tcp6 0 0 :::6379 :::* LISTEN 11470/docker-proxy
tcp6 0 0 :::5644 :::* LISTEN 10929/docker-proxy
tcp6 0 0 :::8080 :::* LISTEN 16133/docker-proxy
tcp6 0 0 :::9200 :::* LISTEN 9838/docker-proxy
tcp6 0 0 :::80 :::* LISTEN 6290/docker-proxy
tcp6 0 0 :::4433 :::* LISTEN 7301/docker-proxy
tcp6 0 0 :::4434 :::* LISTEN 7289/docker-proxy
tcp6 0 0 :::5044 :::* LISTEN 10945/docker-proxy
tcp6 0 0 :::9300 :::* LISTEN 9824/docker-proxy
tcp6 0 0 :::8086 :::* LISTEN 6572/docker-proxy
tcp6 0 0 :::3000 :::* LISTEN 22342/docker-proxy
tcp6 0 0 :::55000 :::* LISTEN 24047/docker-proxy
tcp6 0 0 :::9400 :::* LISTEN 17279/docker-proxy
tcp6 0 0 :::7000 :::* LISTEN 16945/docker-proxy
tcp6 0 0 :::8090 :::* LISTEN 6262/docker-proxy
tcp6 0 0 :::443 :::* LISTEN 6276/docker-proxy
tcp6 0 0 :::9500 :::* LISTEN 17265/docker-proxy
tcp6 0 0 :::9822 :::* LISTEN 7159/docker-proxy
tcp6 0 0 :::3200 :::* LISTEN 18884/docker-proxy
tcp6 0 0 :::9696 :::* LISTEN 11455/docker-proxy
tcp6 0 0 :::9600 :::* LISTEN 10858/docker-proxy
tcp6 0 0 :::5601 :::* LISTEN 13709/docker-proxy
tcp6 0 0 :::514 :::* LISTEN 15335/docker-proxy
tcp6 0 0 :::6050 :::* LISTEN 10915/docker-proxy
tcp6 0 0 :::6051 :::* LISTEN 10902/docker-proxy
tcp6 0 0 :::6052 :::* LISTEN 10886/docker-proxy
tcp6 0 0 :::6053 :::* LISTEN 10872/docker-proxy

**Comment by BustedSec** [comment]
also, I got it temporarily working by just telling ip tables to allow
friggin everything while debugging, then selectively seeing what is needed
and then enabling that ip tables rule. I don't recommend this (security
nightmare) - but if you wanna try it
  112  iptables -F
  113  iptables -I INPUT -j ACCEPT
  114  iptables -F
  115  iptables -X
  116  iptables -t nat -F
  117  iptables -t nat -X
  118  iptables -t mangle -F
  119  iptables -t mangle -X
  120  iptables -P INPUT ACCEPT
  121  iptables -P FORWARD ACCEPT
  122  iptables -P OUTPUT ACCEPT

after a reboot it'll go back to normal - then you add the rules you need.

**Comment by innovate-support** [comment]
Thanks for pointing me in the right direction.  Here's my latest info.

I reviewed the Manager Alerts and can see that the Sensor is trying communicate with the Manager during setup on port 4506, but the Manager is not replying back.  It's pretty handy that this is a packet sniffer so we can see the packets coming in. :-)  So I ran 'sudo soup' on the Manager to get it updated to v2.3.21.  

Based on the recommendations above I read the info found here https://docs.securityonion.net/en/2.3/firewall.html?highlight=firewall and decided to make some firewall edits to see what would happen. I reviewed the firewall settings to see if the Manager was even adding the Sensor's IP to it's configs by running 'sudo so-firewall includedhosts sensor' and 'sudo so-firewall includedhosts minion'.  It wasn't.  So I manually added it with 'sudo so-firewall includehost sensor X.X.X.X' and then ran the setup on the remote Sensor again.  No success.  

I ran 'sudo so-firewall includehost minion X.X.X.X' on the Manager and then ran the setup again on the Sensor and finally got a different result!  It still didn't work but at least we're getting closer.  The new error in the Sensor logs says 'The Salt Master has cached the public key for this node, this salt minion will wait for 10 seconds before attempting to re-authenticate.  Minion failed to authenticate with the master, has the minion key been accepted?' 

So now I'm thinking that if I can manually accept the minion key we may be good to go?  Any know how to do that?  Searching for the answer in these discussion boards didn't result in an answer.

**Reply by BustedSec** [reply]
After running the  sudo so-firewall includehost sensor <ip> command I then I also _edited  /opt/so/saltstack/local/pillar/minions/sensorname.sls (don't do this... just restarted minion and it won't reconnect)_ and changed the IP address from the local IP to the WAN IP.  Then I ran sudo service salt-master restart and I'm seeing traffic from one of my sensors. Working on troubleshooting the other one.

**Reply by BustedSec** [reply]
might wanna confirm 5644 is open (beats) from the forward node to the manager node by scanning your manager node with  nmap -p 5644 managernodeip. Also 5604 and 5605 for salt comms. Might need to install nmap with yum first

**Reply by BustedSec** [reply]
oh! you need to accept the key. Just type salt-key and get a list of all keys then type salt-key -a hostname

**Reply by innovate-support** [reply]
@BustedSec I think we could work together easily.  We seem to think alike.

First nmap scan showed port 5644 not getting through.  So I opened it in edge firewall to make sure it could reach the Manager from the outside, and second scan resulted in a successful hit.

I ran the setup on Sensor again.  It says there's still errors, and the hostname now shows up on the Manager's 'Unaccepted Keys' list when running 'sudo salt-key -l all'.  But at least something's happening!

**Reply by BustedSec** [reply]
If you used the same hostname on the new sensor as the old one you're gonna have a bad day. Just guessing that might be the issue. I've also found that re-running setup on a previously setup box never seems to work.

**Reply by BustedSec** [reply]
During a normal sensor setup - it needs ports 4505, 4506, 5000, and 443 to be open (on the manager). 5644 should also be open once the setup starts. What I think is happening is the script is designed to work on a local network even in a distributed model - and when we introduce NAT - things are breaking  - instead of adding the entries to allow traffic from the WAN address, it's adding the entries to allow traffic through the firewall for the LAN IP addresses. - I've been skirting this issue with these kinda commands on initial setup from the cloud based manager 

with WANIP replaced with your WAN IP. 
sudo iptables --wait -t filter -I INPUT 1 -p tcp --dport 4506 --source WANIP --jump ACCEPT
sudo iptables --wait -t filter -I INPUT 1 -p tcp --dport 4505 --source WANIP --jump ACCEPT
sudo iptables --wait -t filter -I INPUT 1 -p tcp --dport 5000 --source WANIP --jump ACCEPT
sudo iptables --wait -t filter -I INPUT 1 -p tcp --dport 5644--source WANIP --jump ACCEPT

also I added a port forward rule on my router to open up 9300 and 9696 and maps them to the IP of the forward node

**Reply by BustedSec** [reply]
also, if you mess stuff up with salt - you can just delete the key on the manager with the salt-key -d command then kill the salt-minion service (on the forward node) and any associated PIDs. then run salt-minion -l debug and the minion will send over a new key - then you just accept it on the manager. Then kill the debugged salt-minion call and restart the salt-minion service. (on the forward node)

**Reply by innovate-support** [reply]
I manually accepted the unaccepted salt-key using
'sudo salt-key -a sensorname_sensor'

Reran the setup on the Sensor and now am getting new errors.  At least some communication is happening!!
'The SLS minion file is missing on the salt master'.  Looking in the Manager's /opt/so/saltstack/local/pillar/minions/ folder shows only the Manager's sls file.

**Reply by BustedSec** [reply]
if you do a cat sosetup.log | grep error 
what errors do you have?

**Reply by jcmadick** [reply]
Sorry, I'd kind of set this to the side temporarily.  I'll test this approach out this evening.  I thought it had something to do with the cert vs IP address, and my DevOps team isn't terribly helpful at the moment.

**Comment by innovate-support** [comment]
Well, after manually clearing out the Sensors /root/.ssh/known_keys file, then re-running setup on the Sensor again, it asked me to re-login to the Manager during the final setup step to copy the key to the Manager.  After stalling at the '95% Verifying setup' step for a long time, it finally completed (still says there were errors) and a minion.sls file was auto created on the Manager in the /opt/so/saltstack/local/pillar/minions/ folder.  After a reboot of the Sensor it appeared in the Manager's Grid section in the web interface!  Progress!

Now we wait a bit to see if any data actually comes in.  And then I'm gonna delete both Manager and Sensor and start over, because this crap shouldn't be happening this way and I'm not convinced it's really working right.  I need to find out where the broken link is in this chain.

**Reply by innovate-support** [reply]
Perhaps there's some edge firewall issue causing the trouble and a VPN would resolve?  Not sure, but that seems like an unnecessary layer of complexity that I don't want to mess with right now.

**Reply by innovate-support** [reply]
Network traffic is flowing in!  Yay!

**Reply by BustedSec** [reply]
Good job man!

**Comment by jcmadick** [comment]
That worked great!  Thanks you both, I never would have found that direction on my own.

**Reply by BustedSec** [reply]
I'm happy we could help. When you mark this one as complete, maybe gimme the credit :) It'd be cool to have answered one of these.

**Comment by innovate-support** [comment]
I'm happy to give @BustedSec all the credit for answering this.  To close the loop on this one for those who always scroll to the bottom to see what the final answers were, here's my final notes after completely deleting both Manager and Sensor VM's and rebuilding them from scratch. 

**On Edge router firewall**:
MUST add firewall rules that allow ports from whatever public IP that the remote 
sensor will be seen by the Manger.
22
4505
4506
3142
5000
8090
You may just want to allow ALL ports from the remote sensor IP.  

**On Manager**:
MUST add a firewall rule that allows whatever public IP that the remote sensor will be seen as.
Otherwise the remote sensor auth key never gets approved and it’s setup won’t complete.
This is supposed to happen automatically by Manager during the Sensor install, but it doesn’t work.
	$ sudo so-firewall includehost minion X.X.X.X

**On Sensor**:
Do the firewall stuff BEFORE running the setup on the sensor, otherwise it will throw errors, 
the setup fails, you’ll have ghost auth keys that can’t get authenticated unless they’re manually 
approved or deleted.

**After Sensor install and reboot**:
You can check the following on the Manager to confirm the new Sensor was installed successfully.

Check the Grid section on the Manager webpage.  The new Sensor should appear and you should be able to access the Sensor’s Grafana info.

See the Sensor’s accepted key
	$ sudo salt-key	

See the Sensor’s SLS file
	$ sudo ls /opt/so/saltstack/local/pillar/minions/

**Reply by BustedSec** [reply]
one last thing - On Manager:
also might wanna do a 
$ sudo so-firewall includehost sensor X.X.X.X

**Reply by innovate-support** [reply]
after posting my above steps claiming success, no data would come in from the Sensor to the Manager.  It was connected, all statuses showed OK, refreshing Salt and/or Docker didn't affect anything.  The Filebeat logs showed that the Sensor was trying to send stuff to the Manager, but there was a hostname issue.  The Sensor was trying to send to the Manger's private IP, not it's Public one.  So of course the route couldn't be found.  After screwing around with it for a while I gave up and erased everything again.  Something is broken in the way Sensors are communicating with the Manager. Here's the install logs from the Sensor that shows exactly when the failure happens.  I've commented out my Public IP and made it X.X.X.X
[SO sensor setup log.txt](https://github.com/Security-Onion-Solutions/securityonion/files/5890106/SO.sensor.setup.log.txt)

**Reply by innovate-support** [reply]
So after rebuilding the Manager and Sensor again with v2.3.21 on both, the Sensor enrolled on the Manager as intended.  This happened mysteriously since I didn't do anything different than the last 50 times I've tried it.  
But after the required Sensor reboot when the install finished, I don't get any logs from the Sensor to the Manager.  The filebeat.log says it's trying to connect to the Manager's private IP, not the public one that it should.  Anyone know how to resolve that?

2021-01-29T00:08:45.214Z	ERROR	[publisher_pipeline_output]	pipeline/output.go:154	Failed to connect to backoff(async(tcp://so:5644)): dial tcp 10.11.12.155:5644: connect: no route to host

**Reply by BustedSec** [reply]
I don't know if this will work, but try editing /opt/so/saltstack/local/pillar/minions/sensorname.sls changing the IP in there and then restarting stuff.

**Reply by innovate-support** [reply]
the sensor2_sensor.sls file doesn't contain the Manager's IP.  It only has the Manager's hostname.
manager: 'so'

The global.sls file contains a setting for the Manager's IP which currently shows the **private** IP, and if I change it to the **public** IP it does ripple down to the Sensor and the filebeat.log does reflect the change and data magically flows in from the Sensor to the Manager!

Now we reboot and see if the IP change in the global.sls file breaks anything.

**Reply by innovate-support** [reply]
Just as I thought, after a reboot about half the tools break and won't start.  I ran `sudo so-docker-refresh` and it combed through it's list, pulled a few updates, and then everything came back online and so-status shows all OK again.  
`sudo salt \* so.status` shows all OK on both Manager and Sensor.

But...

Although I can log into the web interface, most clicks result in an spinning icon and 504 error.
I am able to open Kibana and review existing logs, but nothing new is coming in from the Sensor.  
The filebeat.log on the Sensor says 
ERROR	[publisher_pipeline_output]	pipeline/output.go:154	Failed to connect to backoff(async(tcp://so:5644)): dial tcp X.X.X.X:5644: connect: connection refused

So I rebooted the Manager again to see what would happen and the only containers that will start are
so-aptcacherng and so-dockerregistry
Running `sudo so-docker-refresh` cause no change this time, and all the containers are stuck in a WAIT_START state.
I am able to manually start each one like `sudo so-filebeat-start`.
Once all the containers are started and show OK the web interface is still spinning and/or giving 504 errors for 'Alerts', 'Hunt', and 'Users', but Kibana seems to be working and after rebooting the Sensor it is able to send logs to the Manager again.
Grafana opens but there's no data populated for either the Manager or the Sensor.

I went ahead and installed a second Sensor VM on another of our servers and it authenticated fine.  Data started flowing into the Manager right away.

Looking in the Sensors /etc/hosts file shows that my Manager's hostname is mapped to it's public IP as desired.  But looking in the /etc/hosts file on the Manager shows only the 127.0.0.1 localhost ip mapped to the hostname.  If I add the public IP to the list so the Manager's IP will map to the public number, it breaks Salt because Salt can't find the Manager if it's not localhost.  

Conclusion: changing to the Manager's public IP in global.sls allowed the Sensors to communicate properly, but broke other functionality such as Grafana, Alerts, and Hunt.  Using Manager's private IP in global.sls will get Manager tools working again, but will break Sensor data flowing in.  So we need a way to define the Manager's public IP on the Sensors.  Their minion.sls files are supposed to take care of this, but there's no static IP setting that I see available to modify.  It uses the Manager's hostname only, which for some reason is resolving to the Manager's private IP instead of it's public IP as desired.  Can I add a public IP in the minion.sls file somehow?  How do I solve this?!

**Reply by horwedja** [reply]
heads up, struggled for ages after applying sudo so-firewall includehost minion X.X.X.X. i then ran **sudo so-firewall apply** and everything worked. Thanks for all the great info!!!!

**Comment by BustedSec** [comment]
This is probably bad advice. Try messing with the hosts file?

**Reply by innovate-support** [reply]
yes, I did mess around with the /etc/hosts file on the Manager and it does affect everything.  If I define the Manager's hostname as anything other than 127.0.0.1 it breaks everything.  It seems that SO relies very heavily on that single setting of the Manager's hostname resolving to localhost.

**Reply by innovate-support** [reply]
One of the annoyances to this whole thing is that a Sensor's entire job is to collect data remotely from the Manager and then feed that data into the Manager.  Yet it's broken, and seemingly non-functional in this current version without some modification that has yet to be discovered.  And to top it off the Salt states all show OK even though this core function of the Sensor is broken and not communicating with the Manager.  Seems like it should flag an error on the Grid if even though the Sensor is Online, if the stream of data can't be sent from Sensor to Manager then it's not OK.

**Comment by BustedSec** [comment]
agreed. In my setup - the Manager is on hosting that has a direct IP on the
internet - but I do see how adding that second NAT into the equation would
mess everything up.

**Reply by innovate-support** [reply]
I am able to give the Manager a public IP directly to the internet, and I was thinking of doing that actually, but that shouldn't be necessary.  I mean SO is a security appliance for pete's sake.

**Reply by jcmadick** [reply]
I'm in the process of rebuilding mine currently.  However, when I successfully registered, my manager was behind a NAT and the sensor was a VM on my workstation, so effectively triple natted.  I could see the connection over 4505 between the two boxes.  I'm tempted, for the sake of expediency, of installing OpenVPN and setting a cron job for regular check-ins.  I managed several old SO 16 boxes that way, was just hoping to avoid it this time around.  Also means no ISO, which isn't necessarily a bad thing since I prefer Ubuntu.

**Reply by innovate-support** [reply]
@weslambert @dougburks @TOoSmOotH @defensivedepth 
I'd like the SO team to chime in on this.  IF they suggest giving the Manager a public IP then I can do that, if they tell me that Sensors aren't gonna work with NAT.  I feel like I've come to the end of my conclusions on this subject and it's an either/or situation.

**Reply by BustedSec** [reply]
jcm - I've tried that - the setup breaks the openvpn tunnel when it disables all other NICs that aren't the monitor interface or main network. I also did a project I called encapsulator that uses IPtables on another machine on a private network between the minion and itself - essentially it uses dnsmasq to distribute DNS\DHCP and gives the manager an IP address then wraps all traffic over the vpn tunnel that is actually on the second box. Getting the VPN tunnel working right is also kinda a PITA.

**Reply by TOoSmOotH** [reply]
@innovate-support When the sensor sends the IP to the manager to open the firewall it is sending what its local IP is. It does not know you are natting it. In theory if you do a site to site VPN with routing, then the manager does not need a public IP address. You would ride that tunnel to the manager over the internet and be fine assuming its an external network device managing the tunnel. What is planned for 2.3.30 is to allow you to set up an openvpn tunnel on the sensor before you run setup and be able to select that interface as your "management" address. It would then send that IP to the manager to open in the firewall and everything would work as it does now and ride the client vpn network vs the normal network at that point. Tracking this with #1139.

**Reply by garthmahns** [reply]
Following on this, as I have the same issue. I have a sensor that set up with "no errors" however the sensor is not attached to the cockpit webpage on grid or anywhere else. So I cannot see alerts or anything else coming from the sensor

**Comment by innovate-support** [comment]
I rebuilt everything again and gave my Manager a public IP.  The Sensors are communicating fine now, but I still had to manually add the so-firewall rules on the Manager BEFORE setting up the Sensors.  The rules are to allow the Sensor's publicly seen IP's to communicate with the Manager.  Of course if your Sensors have changing public IP's because of ISP DHCP configurations you're gonna have fun keeping on top of it.
'sudo so-firewall includehost minion X.X.X.X'
'sudo so-firewall includehost sensor X.X.X.X'

**Reply by BustedSec** [reply]
Workaround Solution for the changing IPs - use a DUC with no-ip to get the IP to resolve to the same name no matter when the IP changes on the minion and then setup a cron job on the manager to do a dig +short hostname.dyndns.com and save it to a variable. Any time the newIP doesn't match the oldIP, run the so-firewall command with the $newIP in a script.

**Reply by BustedSec** [reply]
After the above two commands above, make sure to do a so-firewall apply to apply the changes, btw.

**Comment by sam-acker** [comment]
Just wanted to make it known to anyone reading this thread that this is still a problem in 2.3.61. Big thanks to @innovate-support and @BustedSec for struggling through this and finding a solution! `sudo so-firewall includehost minion X.X.X.X` and `sudo so-firewall includehost sensor X.X.X.X` on the manager node worked for us given that the manager has a public IP and traffic over salt ports are allowed on any edge routers.

**Reply by BustedSec** [reply]
Thanks Sam! - I hope they include this in the offical documentation
